src/driver/xdma_sgdma.c

```diff
@@ -236,22 +236,9 @@ static int xdma_channel_map_pages_to_sgl(struct xdma_io_cb *cb, bool write,
 	return 0;
 }
 
-/* xdma_channel_read_write() -- Read from or write to the device
- *
- * @buf userspace buffer
- * @count number of bytes in the userspace buffer
- * @pos byte-address in device
- * @dir_to_device If !0, a write to the device is performed
- *
- * Iterate over the userspace buffer, taking at most 255 * PAGE_SIZE bytes for
- * each DMA transfer.
- *
- * For each transfer, get the user pages, build a sglist, map, build a
- * descriptor table. submit the transfer. wait for the interrupt handler
- * to wake us on completion.
- */
-ssize_t xdma_channel_read_write(struct xdma_channel *chnl,
-		const char __user *buf, size_t count, loff_t *pos, bool write)
+static ssize_t xdma_channel_read_write(struct xdma_channel *chnl,
+		void *ptr, struct page **pages, unsigned long pages_nr,
+		size_t count, loff_t *pos, bool write)
```

Comment:
This still seems a great deal more messy than it needs to be.

I'd change it so xdma_channel_read_write() takes in an iov_iter.

Then, you don't need a xdma_channel_read_write_buf() as you can very roughly do something like:

```c
struct iov_iter iter;
struct iovec iovec = {
  .iov_base = buf,
  .iov_len = len,
};

iov_iter_init(&iter, WRITE, &iovec, 1, len);
ret = xdma_channel_read_write(chnl, &iter, pos);
```

xdma_channel_read_write() just needs to convert the iter into a sgl and do the (possibly partial) operation. (So the read_iter/write_iter functions will need to call it in a loop and it will do one page of SGLs per iteration, or some such).

The direction is then  encoded in the iter and we don't need the bool. Gets rid a lot of duplicate mess in the read_iter and write_iter functions.

For converting the iter to the SGL, you could use a trick similar to this:

https://elixir.bootlin.com/linux/latest/source/block/bio.c#L1012

Where you allocate an SGL, then fill the pages into the end of the SGL and then put the pages in the correct page pointers for the SGL.  This saves an extra allocation.

Reply:
Hmm I like the idea of converting the buf into an iov_iter, that will make it cleaner.

But I'm not sure I fully understand your second suggestion of how to convert the iov_iter into the sgl:
1. Is the extra allocation you're trying to eliminate the one from `iov_iter_get_pages_alloc`? On that link you posted, it can call `iov_iter_get_pages` because the bio already has the `struct page`s in the struct.
2. I think that moving the `iov_iter_get_pages(_alloc)`/`iov_iter_advance` from `hermes_read_iter`/`hermes_write_iter` to `xdma_channel_map_pages_to_sgl` would accomplish what you're suggesting (plus some error checking and ensuring all pages were processed). Is that right?

Reply:
The bio stores the pages in a bio_vec which is analogous to a scatterlist, so it's the same issue.

So, to do it the naive way you have to allocate memory for the list of pages and also memory for the SGL itself roughly:

```c
struct page **pages;
struct sg_table sgt;

iov_iter_get_pages_alloc(iter, &pages, ...);
sg_alloc_table_from_pages(&sgt, pages, ...);
kvfree(pages);

... do something with the sgl ...
iov_iter_advance(iter, ...);

for_each_sg_page(...)
    put_page(pg);
sg_free_table(&sgt);
```

But, we don't actually need to allocate memory for both the page array and the sgl, we can use the same memory, like they do in __bio_iov_iter_get_pages()...

```c
#define PAGE_PTRS_PER_SGL (sizeof(struct scatterlist) / sizeof(struct page *))
struct page **pages;
struct sg_table sgt;
ssize_t size, left;
size_t offset;
int i;

sg_alloc_table(&sgt, SG_MAX_SINGLE_ALLOC, ...);
pages = (struct page **)sgt.sgl;
pages += SG_MAX_SINGLE_ALLOC * (PAGE_PTRS_PER_SGL - 1);

size = iov_iter_get_pages(iter, LONG_MAX, SG_MAX_SINGLE_ALLOC, &offset);
for (left = size, i = 0; left > 0; left -= len, i++) {
		len = min_t(size_t, PAGE_SIZE - offset, left);
		sg_set_page(&sgt.sgl[i], pages[i], len, offset);
                offset = 0;
}

... do something with the sgl ...

iov_iter_advance(iter, size);
for_each_sg_page(...)
    put_page(pg);
sg_free_table(&sgt);

```

So for each run of this function, we only allocate one page of memory (which is the fastest allocation) and can transfer up to about 128 pages (or 512KB). Or, even better,  if we allocate the sg table outside of the loop we only allocate and free one page of memory for any size transfer. Roughly:

```c
#define PAGE_PTRS_PER_SGL (sizeof(struct scatterlist) / sizeof(struct page *))
struct page **pages;
struct sg_table sgt;
ssize_t size, left;
size_t offset;
int i;

sg_alloc_table(&sgt, SG_MAX_SINGLE_ALLOC, ...);

while (iov_iter_count(iter)) {
        pages = (struct page **)sgt.sgl;
        pages += SG_MAX_SINGLE_ALLOC * (PAGE_PTRS_PER_SGL - 1);

        size = iov_iter_get_pages(iter, LONG_MAX, SG_MAX_SINGLE_ALLOC, &offset);
        for (left = size, i = 0; left > 0; left -= len, i++) {
        		len = min_t(size_t, PAGE_SIZE - offset, left);
        		sg_set_page(&sgt.sgl[i], pages[i], len, offset);
                        offset = 0;
        }

        ... do something with the sgl ...

        iov_iter_advance(iter, size);
        for_each_sg_page(...)
                   put_page(pg);
}
sg_free_table(&sgt);

```

Reply:
Got it, thanks!!!